File: 백준/Silver/2304. 창고 다각형/창고 다각형.py
def solution(cols: int, rows: int):
    answer = 0
    arr_left = [[0] * (cols + 2) for _ in range(rows + 2)]
    arr_right = [[0] * (cols + 2) for _ in range(rows + 2)]

    for pillar in pillars:
        arr_left[pillar[1]][pillar[0]] += 1
        arr_right[pillar[1]][pillar[0]] += 1

    for y in range(rows, 0, -1):
        for x in range(1, cols + 1):
            if arr_left[y][x - 1] == 1 or arr_left[y + 1][x] == 1:
                arr_left[y][x] = 1

    for y in range(rows, 0, -1):
        for x in range(cols, 0, -1):
            if arr_right[y][x + 1] == 1 or arr_right[y + 1][x] == 1:
                arr_right[y][x] = 1

            if arr_left[y][x] == 1 and arr_right[y][x] == 1:
                answer += 1

    print(answer)

n = int(input())
pillars = [[0, 0]] + [list(map(int, input().split())) for _ in range(n)]

rows = max(pillar[1] for pillar in pillars)
cols = max(pillar[0] for pillar in pillars)

solution(cols, rows)

File: 백준/Silver/2304. 창고 다각형/창고 다각형.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] pillars = new int[n][2];
        for (int i = 0; i < n; i++) {
            String[] num = br.readLine().split(" ");
            pillars[i][0] = Integer.parseInt(num[0]);
            pillars[i][1] = Integer.parseInt(num[1]);
        }

        System.out.println(solution(n, pillars));
    }

    static int solution(int n, int[][] pillars) {
        Arrays.sort(pillars, Comparator.comparingInt(pillar -> pillar[0]));
        int max = getMax(pillars);
        int answer = 0;

        // 정방향
        int[] prefixA = new int[]{pillars[0][0], pillars[0][1]};
        for (int i = 1; i < n; i++) {
            answer += (pillars[i][0] - prefixA[0]) * prefixA[1];
            prefixA[0] = pillars[i][0];
            prefixA[1] = Math.max(prefixA[1], pillars[i][1]);
            if (prefixA[1] == max) {
                break;
            }
        }

        // 역방향
        int[] prefixB = new int[]{pillars[n - 1][0], pillars[n - 1][1]};
        for (int i = n - 2; i >= 0; i--) {
            answer += (prefixB[0] - pillars[i][0]) * prefixB[1];
            prefixB[0] = pillars[i][0];
            prefixB[1] = Math.max(prefixB[1], pillars[i][1]);
            if (prefixB[1] == max) {
                break;
            }
        }

        // 중간
        answer += (prefixB[0] - prefixA[0] + 1) * max;

        return answer;
    }

    static int getMax(int[][] arr) {
        int max = Integer.MIN_VALUE;
        for (int[] ints : arr) {
            if (ints[1] > max) {
                max = ints[1];
            }
        }
        return max;
    }
}